README.md
# Ansible Collection - ridwanbejo.sentry

Documentation for the collection.


## A. Requirements

lorem ipsum sit dolor amet


## B. How to install

lorem ipsum sit dolor amet


## C. How to use the modules

lorem ipsum sit dolor amet


## D. How to test it locally

lorem ipsum sit dolor amet


## E. How to contribute

lorem ipsum sit dolor amet



CHANGELOG.md
CHANEGELOG.md


test.sh
#!/bin/bash

ANSIBLE_GIT_PATH=/home/ridwansense/Documents/ridwanbejo/repository/ansible

LOCAL_ANSIBLE_COLLECTIONS_PATH=/home/ridwansense/.ansible/collections/ansible_collections

. ${ANSIBLE_GIT_PATH}/venv/bin/activate

. ${ANSIBLE_GIT_PATH}/hacking/env-setup

cd ${LOCAL_ANSIBLE_COLLECTIONS_PATH}

ansible-test sanity --test pep8


plugins/modules/sentry_project.py
#!/usr/bin/python

# Copyright: (c) 2022, Ridwan Fadjar Septian <ridwanbejo@gmail.com>
# GNU General Public License v3.0+ (see COPYING or https://www.gnu.org/licenses/gpl-3.0.txt)


from __future__ import absolute_import, division, print_function


__metaclass__ = type



DOCUMENTATION = r"""
module: sentry_project
short_description: Manage projects in Sentry. Currently supported operation are create, update and delete.
author:
    - "ridwanbejo (@ridwanbejo)"
description:
  - Based on Sentry API documentation (https://docs.sentry.io/api/), this module will help you to manage your projects
options:
  sentry_host:
    description:
    - Target hostname of Sentry
    type: str
    default: true
    version_added: 1.0.0
  sentry_token:
    description:
    - Token which generated in Sentry by administrator. This token is located under "Settings > Internal Integration"
    type: str
    default: true
    version_added: 1.0.0
  organization_slug:
    description:
    - Slug of the organization
    type: str
    default: true
    version_added: 1.0.0
  team_slug:
    description:
    - slug of the team
    type: str
    default: false
    version_added: 1.0.0
  project_slug:
    description:
    - slug for the chosen project
    type: str
    default: false
    version_added: 1.0.0
  name:
    description:
    - new name for the project
    type: str
    default: false
    version_added: 1.0.0
  slug:
    description:
    - new slug for the project
    type: str
    default: false
    version_added: 1.0.0
  platform:
    description:
    - Platform for the chosen project
    type: str
    default: false
    version_added: 1.0.0
  is_bookmarked:
    description:
    - Bookmark the project
    type: bool
    default: false
    version_added: 1.0.0
  state:
    description:
      - Perform operation to create, update and delete project in Sentry
    default: 'present'
    choices: [present', 'absent']
    type: str
requirements:
    - "python >= 3.8.10"
    - "ansible >= 2.12.1"
    - "requests >= 2.26.0"
"""

EXAMPLES = r"""
# Create new project in Sentry
- name: Test Sentry Project module - create project
    ridwanbejo.sentry.sentry_project:
      sentry_host: "http://localhost:9000"
      sentry_token: "8702e9c2d5224b60b24d2f7a9fa486f0eaaee9748a0e4acda3ea8febdc790093"
      name: 'Bonjour'
      slug: 'bonjour'
      organization_slug: 'sentry'
      team_slug: 'sentry'
      state: present

# Update project which has slug bonjour in Sentry with new slug and name
- name: Test Sentry Project module - update project
    ridwanbejo.sentry.sentry_project:
      sentry_host: "http://localhost:9000"
      sentry_token: "8702e9c2d5224b60b24d2f7a9fa486f0eaaee9748a0e4acda3ea8febdc790093"
      organization_slug: 'sentry'
      project_slug: 'bonjour'
      name: 'Buongiorno'
      slug: 'bonjour-monsieur'
      platform: "python"
      is_bookmarked: true
      state: present

# Delete project which has slug bonjour-monsieur
- name: Test Sentry Project module - delete project
    ridwanbejo.sentry.sentry_project:
      sentry_host: "http://localhost:9000"
      sentry_token: "8702e9c2d5224b60b24d2f7a9fa486f0eaaee9748a0e4acda3ea8febdc790093"
      organization_slug: 'sentry'
      project_slug: 'bonjour-monsieur'
      state: absent
"""

RETURN = r"""
"""

import requests
import json

from ansible.module_utils.basic import AnsibleModule
from ansible_collections.ridwanbejo.sentry.plugins.module_utils.sentry_api import SentryApi


def run_module():
    module_args = dict(
        sentry_host=dict(type='str', require=True),
        sentry_token=dict(type='str', require=True),
        organization_slug=dict(type='str', required=True),
        team_slug=dict(type='str', required=False),
        project_slug=dict(type='str', required=False),
        name=dict(type='str', required=False),
        slug=dict(type='str', required=False),
        platform=dict(type='str', required=False),
        is_bookmarked=dict(type='bool', required=False),
        state=dict(
            default="present", 
            choices=['present', 'absent'],  
            type='str')
    )

    module = AnsibleModule(
        argument_spec=module_args,
        supports_check_mode=True
    )

    if module.check_mode:
        module.exit_json(dict(
            changed=False,
            message='Check mode success!'
        ))

    result = dict(
    )

    sentry_api = SentryApi(module, module.params['sentry_host'], module.params['sentry_token'])

    # a. if state is present then check the existence of project
    if module.params['state'] == "present":

        # a.1. if the project is not exist then create new project
        retrieve_requests = sentry_api.retrieve_project(
            module.params['organization_slug'],
            module.params['project_slug']
        )

        if retrieve_requests['status_code'] == 200:
            result = sentry_api.update_project(
                module.params['organization_slug'],
                module.params['project_slug'],
                module.params['team_slug'],
                module.params['name'],
                module.params['slug'],
                module.params['platform'],
                module.params['is_bookmarked']
            )

            if result['status_code'] != 200:
                module.fail_json(msg=result)

        # a.2. if the project is exist before then update the project
        elif retrieve_requests['status_code'] == 404:

            result = sentry_api.create_project(
                module.params['organization_slug'],
                module.params['team_slug'],
                module.params['name'],
                module.params['slug']
            )

            if result['status_code'] != 201:
                module.fail_json(msg=result)

    # b. if state is absent then delete the project
    elif module.params['state'] == "absent":
        result = sentry_api.delete_project(
            module.params['organization_slug'],
            module.params['project_slug']
        )

        if result['status_code'] != 204:
            module.fail_json(msg=result)

    module.exit_json(**result)


def main():
    run_module()


if __name__ == '__main__':
    main()


plugins/module_utils/sentry_api.py
#!/usr/bin/python

import requests
import json


class SentryApi(object):
	CREATE_PROJECT_URL = "/api/0/teams/{organization_slug}/{team_slug}/projects/"
	RETRIEVE_PROJECT_URL = "/api/0/projects/{organization_slug}/{project_slug}/"
	UPDATE_PROJECT_URL = "/api/0/projects/{organization_slug}/{project_slug}/"
	DELETE_PROJECT_URL = "/api/0/projects/{organization_slug}/{project_slug}/"

	CREATE_TEAM_URL = "/api/0/organizations/{organization_slug}/teams/"
	RETRIEVE_TEAM_URL = "/api/0/teams/{organization_slug}/{team_slug}/"
	UPDATE_TEAM_URL = "/api/0/teams/{organization_slug}/{team_slug}/"
	DELETE_TEAM_URL = "/api/0/teams/{organization_slug}/{team_slug}/"

	RETRIEVE_ORGANIZATION_URL = "/api/0/organizations/{organization_slug}/"
	UPDATE_ORGANIZATION_URL = "/api/0/organizations/{organization_slug}/"

	CREATE_CLIENT_KEY_URL = "/api/0/projects/{organization_slug}/{project_slug}/keys/"
	UPDATE_CLIENT_KEY_URL = "/api/0/projects/{organization_slug}/{project_slug}/keys/{client_key}/"
	DELETE_CLIENT_KEY_URL = "/api/0/projects/{organization_slug}/{project_slug}/keys/{client_key}/"

	CREATE_SERVICE_HOOK_URL = "/api/0/projects/{organization_slug}/{project_slug}/hooks/"
	UPDATE_SERVICE_HOOK_URL = "/api/0/projects/{organization_slug}/{project_slug}/hooks/{hook_id}/"
	DELETE_SERVICE_HOOK_URL = "/api/0/projects/{organization_slug}/{project_slug}/hooks/{hook_id}/"

	def __init__(self, module, host, token):
		self.module = module
		self.host = host

		self.headers = {
			'Authorization': 'Bearer '+token, 
			'Content-Type': 'application/json'
		}

		self.result = dict(
			message=''
		)

	def build_url(self, url):
		return self.host+url

	def get_url(self, task):
		if task == 'create-project':
			return self.build_url(self.CREATE_PROJECT_URL.format(
						organization_slug=self.organization_slug,
						team_slug=self.team_slug
					))
		elif task == 'retrieve-project':
			return self.build_url(self.RETRIEVE_PROJECT_URL.format(
						organization_slug=self.organization_slug,
						project_slug=self.project_slug
					))
		elif task == 'update-project':
			return self.build_url(self.UPDATE_PROJECT_URL.format(
						organization_slug=self.organization_slug,
						project_slug=self.project_slug
					))
		elif task == 'delete-project':
			return self.build_url(self.DELETE_PROJECT_URL.format(
						organization_slug=self.organization_slug,
						project_slug=self.project_slug
					))
		elif task == 'create-team':
			return self.build_url(self.CREATE_TEAM_URL.format(
						organization_slug=self.organization_slug
					))
		elif task == 'retrieve-team':
			return self.build_url(self.RETRIEVE_TEAM_URL.format(
						organization_slug=self.organization_slug,
						team_slug=self.team_slug
					))
		elif task == 'update-team':
			return self.build_url(self.UPDATE_TEAM_URL.format(
						organization_slug=self.organization_slug,
						team_slug=self.team_slug
					))
		elif task == 'delete-team':
			return self.build_url(self.DELETE_TEAM_URL.format(
						organization_slug=self.organization_slug,
						team_slug=self.team_slug
					))
		elif task == 'retrieve-organization':
			return self.build_url(self.RETRIEVE_ORGANIZATION_URL.format(
						organization_slug=self.organization_slug
					))
		elif task == 'update-organization':
			return self.build_url(self.UPDATE_ORGANIZATION_URL.format(
						organization_slug=self.organization_slug
					))
		elif task == 'create-client-key':
			return self.build_url(self.CREATE_CLIENT_KEY_URL.format(
						organization_slug=self.organization_slug,
						project_slug=self.project_slug
					))
		elif task == 'update-client-key':
			return self.build_url(self.UPDATE_CLIENT_KEY_URL.format(
						organization_slug=self.organization_slug,
						project_slug=self.project_slug,
						client_key=self.client_key
					))
		elif task == 'delete-client-key':
			return self.build_url(self.DELETE_CLIENT_KEY_URL.format(
						organization_slug=self.organization_slug,
						project_slug=self.project_slug,
						client_key=self.client_key
					))
		elif task == 'create-service-hook':
			return self.build_url(self.CREATE_SERVICE_HOOK_URL.format(
						organization_slug=self.organization_slug,
						project_slug=self.project_slug
					))
		elif task == 'update-service-hook':
			return self.build_url(self.UPDATE_SERVICE_HOOK_URL.format(
						organization_slug=self.organization_slug,
						project_slug=self.project_slug,
						hook_id=self.hook_id
					))
		elif task == 'delete-service-hook':
			return self.build_url(self.DELETE_SERVICE_HOOK_URL.format(
						organization_slug=self.organization_slug,
						project_slug=self.project_slug,
						hook_id=self.hook_id
					))

	def create_project(self, organization_slug, team_slug, name, slug):
		self.organization_slug = organization_slug
		self.team_slug = team_slug
		result = self.result

		payload = {
	        'name': name,
	        'slug': slug
	    }

		create_project_url = self.get_url('create-project')

		create_requests = requests.post(create_project_url, data=json.dumps(payload), headers=self.headers)

		result['url'] = create_project_url
		result['status_code'] = create_requests.status_code
		result['changed'] = True
		result['failed'] = False
		result['message'] = "Project has been created"

		if create_requests.status_code != 201:
			result['changed'] = False
			result['failed'] = True
			result['message'] = "Can't create project"

		result['response'] = create_requests.json()

		return result

	def retrieve_project(self, organization_slug, project_slug):
		self.organization_slug = organization_slug
		self.project_slug = project_slug

		result = self.result

		retrieve_project_url = self.get_url('retrieve-project')

		retrieve_requests = requests.get(retrieve_project_url, headers=self.headers)

		result['url'] = retrieve_project_url
		result['status_code'] = retrieve_requests.status_code
		result['changed'] = False
		result['failed'] = False
		result['message'] = "Project is available"

		if retrieve_requests.status_code != 200:
			result['failed'] = True
			result['message'] = "Can't retrieve project"

		result['response'] = retrieve_requests.json()

		return result

	def update_project(self, organization_slug, project_slug, team_slug, name, slug, platform, is_bookmarked):
		self.organization_slug = organization_slug
		self.project_slug = project_slug

		result = self.result

		payload = {
		    'name': name,
		    'slug': slug,
		    'team_slug': team_slug,
		    'platform': platform,
		    'is_bookmarked': is_bookmarked
		}

		update_project_url = self.get_url('update-project')

		update_requests = requests.put(update_project_url, data=json.dumps(payload), headers=self.headers)

		result['url'] = update_project_url
		result['status_code'] = update_requests.status_code
		result['changed'] = True
		result['failed'] = False
		result['message'] = "Project has been updated"
		
		if update_requests.status_code != 200:
			result['changed'] = False
			result['failed'] = True
			result['message'] = "Can't update orihect"

		result['response'] = update_requests.json()

		return result

	def delete_project(self, organization_slug, project_slug):
		self.organization_slug = organization_slug
		self.project_slug = project_slug
		result = self.result

		delete_project_url = self.get_url('delete-project')

		delete_requests = requests.delete(delete_project_url, headers=self.headers)

		result['changed'] = True
		result['failed'] = False
		result['message'] = "Project has been deleted"
		result['response'] = {
			"detail": "Success"
		}

		if delete_requests.status_code != 204:
			result['changed'] = False
			result['failed'] = True
			result['message'] = "Can't delete project"
			result['response'] = delete_requests.json()

		return result

	def create_team(self, organization_slug, name, slug):
		self.organization_slug = organization_slug
		result = self.result

		payload = {
	        'name': name,
	        'slug': slug
	    }

		create_team_url = self.get_url('create-team')

		create_requests = requests.post(create_team_url, data=json.dumps(payload), headers=self.headers)

		result['url'] = create_team_url
		result['status_code'] = create_requests.status_code

		result['changed'] = True
		result['failed'] = False
		result['message'] = "Team has been created"
		
		if create_requests.status_code != 201:
			result['changed'] = False
			result['failed'] = True
			result['message'] = "Can't create team"

		result['response'] = create_requests.json()
		
		return result

	def retrieve_team(self, organization_slug, team_slug):
		self.organization_slug = organization_slug
		self.team_slug = team_slug

		result = self.result

		retrieve_team_url = self.get_url('retrieve-team')

		retrieve_requests = requests.get(retrieve_team_url, headers=self.headers)

		result['url'] = retrieve_team_url
		result['status_code'] = retrieve_requests.status_code
		
		result['changed'] = False
		result['failed'] = False
		result['message'] = "Team is available"
		
		if retrieve_requests.status_code != 200:
			result['failed'] = True
			result['message'] = "Can't retrieve team"

		result['response'] = retrieve_requests.json()

		return result

	def update_team(self, organization_slug, team_slug, name, slug):
		self.organization_slug = organization_slug
		self.team_slug = team_slug

		result = self.result

		payload = {
		    'name': name,
		    'slug': slug
		}

		update_team_url = self.get_url('update-team')

		update_requests = requests.put(update_team_url, data=json.dumps(payload), headers=self.headers)

		result['url'] = update_team_url
		result['status_code'] = update_requests.status_code

		result['changed'] = True
		result['failed'] = False
		result['message'] = "Team has been updated"

		if update_requests.status_code != 200:
			result['changed'] = False
			result['failed'] = True
			result['message'] = "Can't update team"

		result['response'] = update_requests.json()

		return result

	def delete_team(self, organization_slug, team_slug):
		self.organization_slug = organization_slug
		self.team_slug = team_slug
		result = self.result

		delete_team_url = self.get_url('delete-team')

		delete_requests = requests.delete(delete_team_url, headers=self.headers)

		result['url'] = delete_team_url
		result['status_code'] = delete_requests.status_code

		result['changed'] = True
		result['failed'] = False
		result['message'] = "Team has been deleted"
		result['response'] = {
			"detail": "Success"
		}

		if delete_requests.status_code != 204:
			result['changed'] = False
			result['failed'] = True
			result['message'] = "Can't delete team"
			result['response'] = delete_requests.json()

		return result

	def retrieve_organization(self, organization_slug):
		self.organization_slug = organization_slug

		result = self.result

		retrieve_organization_url = self.get_url('retrieve-organization')

		retrieve_requests = requests.get(retrieve_organization_url, headers=self.headers)

		result['url'] = retrieve_organization_url
		result['status_code'] = retrieve_requests.status_code

		result['changed'] = False
		result['failed'] = False
		result['message'] = "Organization is available"

		if retrieve_requests.status_code != 200:
			result['failed'] = True
			result['message'] = "Can't retrieve the organization"

		result['response'] = retrieve_requests.json()

		return result

	def update_organization(self, organization_slug, name, slug):
		self.organization_slug = organization_slug

		result = self.result

		payload = {
		    'name': name,
		    'slug': slug
		}

		update_organization_url = self.get_url('update-organization')

		update_requests = requests.put(update_organization_url, data=json.dumps(payload), headers=self.headers)

		result['url'] = update_organization_url
		result['status_code'] = update_requests.status_code

		result['changed'] = True
		result['failed'] = False
		result['message'] = "Organization has been updated"

		if update_requests.status_code != 200:
			result['changed'] = False
			result['failed'] = True
			result['message'] = "Can't update organization"

		result['response'] = update_requests.json()

		return result

	def create_client_key(self, organization_slug, project_slug, name):
		self.organization_slug = organization_slug
		self.project_slug = project_slug
		result = self.result

		payload = {
	        'name': name
	    }

		create_client_key_url = self.get_url('create-client-key')

		create_requests = requests.post(create_client_key_url, data=json.dumps(payload), headers=self.headers)

		result['message'] = "Project Client Key has been created"
		result['url'] = create_client_key_url
		result['response'] = create_requests.json()
		result['status_code'] = create_requests.status_code

		return result

	def update_client_key(self, organization_slug, project_slug, client_key, name, is_active):
		self.organization_slug = organization_slug
		self.project_slug = project_slug
		self.client_key = client_key

		result = self.result

		payload = {
		   'name': name,
		   'isActive': is_active
		}

		update_client_key_url = self.get_url('update-client-key')

		update_requests = requests.put(update_client_key_url, data=json.dumps(payload), headers=self.headers)

		result['changed'] = True
		result['message'] = "Project Client Key has been updated"
		result['url'] = update_client_key_url
		result['status_code'] = update_requests.status_code
		result['response'] = update_requests.json()

		return result

	def delete_client_key(self, organization_slug, project_slug, client_key):
		self.organization_slug = organization_slug
		self.project_slug = project_slug
		self.client_key = client_key

		result = self.result

		delete_client_key_url = self.get_url('delete-client-key')

		delete_requests = requests.delete(delete_client_key_url, headers=self.headers)

		result['changed'] = True
		result['message'] = "Project Client Key has been deleted"
		result['url'] = delete_client_key_url
		result['status_code'] = delete_requests.status_code

		if delete_requests.status_code != 204:
			result['response'] = delete_requests.json()

		return result

	def create_service_hook(self, organization_slug, project_slug, hook_url, hook_events):
		self.organization_slug = organization_slug
		self.project_slug = project_slug

		result = self.result

		payload = {
	        'url': hook_url,
	        'events': hook_events
	    }

		create_service_hook_url = self.get_url('create-service-hook')

		create_requests = requests.post(create_service_hook_url, data=json.dumps(payload), headers=self.headers)

		result['message'] = "Project Service Hook has been created"
		result['url'] = create_service_hook_url
		result['response'] = create_requests.json()
		result['status_code'] = create_requests.status_code

		return result

	def update_service_hook(self, organization_slug, project_slug, hook_id, hook_url, hook_events):
		self.organization_slug = organization_slug
		self.project_slug = project_slug
		self.hook_id = hook_id

		result = self.result

		payload = {
	        'url': hook_url,
	        'events': hook_events
	    }

		update_service_hook_url = self.get_url('update-service-hook')

		update_requests = requests.put(update_service_hook_url, data=json.dumps(payload), headers=self.headers)

		result['changed'] = True
		result['message'] = "Project Service Hook has been updated"
		result['url'] = update_service_hook_url
		result['response'] = update_requests.json()
		result['status_code'] = update_requests.status_code

		return result

	def delete_service_hook(self, organization_slug, project_slug, hook_id):
		self.organization_slug = organization_slug
		self.project_slug = project_slug
		self.hook_id = hook_id

		result = self.result

		delete_service_hook_url = self.get_url('delete-service-hook')

		delete_requests = requests.delete(delete_service_hook_url, headers=self.headers)

		result['changed'] = True
		result['message'] = "Project Service Hook has been deleted"
		result['url'] = delete_service_hook_url
		result['status_code'] = delete_requests.status_code

		if delete_requests.status_code != 204:
			result['response'] = delete_requests.json()

		return result


plugins/modules/sentry_organization.py
#!/usr/bin/python

# Copyright: (c) 2022, Ridwan Fadjar Septian <ridwanbejo@gmail.com>
# GNU General Public License v3.0+ (see COPYING or https://www.gnu.org/licenses/gpl-3.0.txt)


from __future__ import absolute_import, division, print_function


__metaclass__ = type



DOCUMENTATION = r"""
module: sentry_organization
short_description: Manage organization information in Sentry. Currently supported operation is update only.
author:
    - "ridwanbejo (@ridwanbejo)"
description:
  - Based on Sentry API documentation (https://docs.sentry.io/api/), this module will help you to update your organization information
options:
  sentry_host:
    description:
    - Target hostname of Sentry
    type: str
    default: true
    version_added: 1.0.0
  sentry_token:
    description:
    - Token which generated in Sentry by administrator. This token is located under "Settings > Internal Integration"
    type: str
    default: true
    version_added: 1.0.0
  organization_slug:
    description:
    - Slug of organization which need to be updated
    type: str
    default: true
    version_added: 1.0.0
  name:
    description:
    - name for the organization
    type: str
    default: false
    version_added: 1.0.0
  slug:
    description:
    - slug for the organization
    type: str
    default: false
    version_added: 1.0.0
  state:
    description:
      - Perform operation to update organization information. This state option is provided to handle another operation from Sentry API
    default: 'present'
    choices: [present']
    type: str
requirements:
    - "python >= 3.8.10"
    - "ansible >= 2.12.1"
    - "requests >= 2.26.0"
"""

EXAMPLES = r"""
# Update default name and slug to new organization name and slug
- name: Update organization in Sentry
    ridwanbejo.sentry.sentry_organization:
      sentry_host: "http://localhost:9000"
      sentry_token: "8702e9c2d5224b60b24d2f7a9fa486f0eaaee9748a0e4acda3ea8febdc790093"
      name: 'My Great Company'
      slug: 'my-great-company'
      organization_slug: 'sentry'
      state: present
"""

RETURN = r"""
msg:
  description:
  - If a change was successfully applied, it will return the updated object, otherwise returns response with status code 40x.
  returned: always
  type: complex
  contains:
     changed:
       description: If update process succeed, changed will be true otherwise false
       returned:always
       type: bool
     failed:
       description: failure status of this module execution
       returned: always
       type: bool
     message:
       description: Additional message along yield by this module
       returned: always
       type: str
       sample: Organization has been updated
     response:
       description: Response from organization related endpoints. Will vary based on the I(api_version) and I(kind). Check more response detail at https://docs.sentry.io/api/organizations/update-an-organization/
       returned: always
       type: complex
     status_code:
       description: Response status code
       returned: always
       type: int
       sample: 200
     url:
       description: requested URL from this module
       returned: always
       type: str
       sample: http://localhost:9000/api/0/organizations/my-great-company/
"""

import requests
import json

from ansible.module_utils.basic import AnsibleModule
from ansible_collections.ridwanbejo.sentry.plugins.module_utils.sentry_api import SentryApi


def run_module():
    module_args = dict(
        sentry_host=dict(type='str', require=True),
        sentry_token=dict(type='str', require=True),
        organization_slug=dict(type='str', required=True),
        name=dict(type='str', required=False),
        slug=dict(type='str', required=False),
        state=dict(
            default="present", 
            choices=['present'],  
            type='str')
    )

    module = AnsibleModule(
        argument_spec=module_args,
        supports_check_mode=True
    )

    if module.check_mode:
        module.exit_json(dict(
            changed=False,
            message='Check mode success!'
        ))

    result = dict(
    )

    sentry_api = SentryApi(module, module.params['sentry_host'], module.params['sentry_token'])

    # a. if state is present then check the existence of organization
    if module.params['state'] == "present":

        # a.1. if the organization is not exist then create new organization
        retrieve_requests = sentry_api.retrieve_organization(
            module.params['organization_slug']
        )

        if retrieve_requests['status_code'] == 200:
            result = sentry_api.update_organization(
                module.params['organization_slug'],
                module.params['name'],
                module.params['slug'],
            )

            if result['status_code'] != 200:
                module.fail_json(msg=result)
        else:
            result = retrieve_requests
            module.fail_json(msg=result)

    module.exit_json(**result)


def main():
    run_module()


if __name__ == '__main__':
    main()


plugins/modules/sentry_team.py
#!/usr/bin/python

# Copyright: (c) 2022, Ridwan Fadjar Septian <ridwanbejo@gmail.com>
# GNU General Public License v3.0+ (see COPYING or https://www.gnu.org/licenses/gpl-3.0.txt)


from __future__ import absolute_import, division, print_function


__metaclass__ = type



DOCUMENTATION = r"""
module: sentry_team
short_description: Manage teams in Sentry. Currently supported operation are create, update and delete.
author:
    - "ridwanbejo (@ridwanbejo)"
description:
  - Based on Sentry API documentation (https://docs.sentry.io/api/), this module will help you to manage your teams
options:
  sentry_host:
    description:
    - Target hostname of Sentry
    type: str
    default: true
    version_added: 1.0.0
  sentry_token:
    description:
    - Token which generated in Sentry by administrator. This token is located under "Settings > Internal Integration"
    type: str
    default: true
    version_added: 1.0.0
  organization_slug:
    description:
    - Slug of the organization
    type: str
    default: true
    version_added: 1.0.0
  team_slug:
    description:
    - slug of the team
    type: str
    default: false
    version_added: 1.0.0
  name:
    description:
    - new name for the team
    type: str
    default: false
    version_added: 1.0.0
  slug:
    description:
    - new slug for the team
    type: str
    default: false
    version_added: 1.0.0
  state:
    description:
      - Perform operation to create, update and delete project in Sentry
    default: 'present'
    choices: [present', 'absent']
    type: str
requirements:
    - "python >= 3.8.10"
    - "ansible >= 2.12.1"
    - "requests >= 2.26.0"
"""

EXAMPLES = r"""
# Create new team in Sentry
- name: Test Sentry Team module - create team
    ridwanbejo.sentry.sentry_team:
      sentry_host: "http://localhost:9000"
      sentry_token: "8702e9c2d5224b60b24d2f7a9fa486f0eaaee9748a0e4acda3ea8febdc790093"
      name: 'Backend Team'
      slug: 'backend-team'
      organization_slug: 'sentry'
      state: present

# Update team which has slug backend-team in Sentry with new slug and name
- name: Test Sentry Team module - update team
    ridwanbejo.sentry.sentry_team:
      sentry_host: "http://localhost:9000"
      sentry_token: "8702e9c2d5224b60b24d2f7a9fa486f0eaaee9748a0e4acda3ea8febdc790093"
      name: 'Backend Developer Team'
      slug: 'backend-dev-team'
      organization_slug: 'sentry'
      team_slug: 'backend-team'
      state: present

# Delete team which has slug backend-dev-team
- name: Test Sentry Team module - delete team
    ridwanbejo.sentry.sentry_team:
      sentry_host: "http://localhost:9000"
      sentry_token: "8702e9c2d5224b60b24d2f7a9fa486f0eaaee9748a0e4acda3ea8febdc790093"
      organization_slug: 'sentry'
      team_slug: 'backend-dev-team'
      state: absent
"""

RETURN = r"""
msg:
  description:
  - If a change was successfully applied, it will return success response with status 20x, otherwise returns response with status code 40x.
  returned: always
  type: complex
  contains:
     changed:
       description: If process succeed, changed will be true otherwise false
       returned:always
       type: bool
     failed:
       description: failure status of this module execution
       returned: always
       type: bool
     message:
       description: Additional message along yield by this module
       returned: always
       type: str
       sample: Team has been created
     response:
       description: Response from team related endpoints. Will vary based on the I(api_version) and I(kind). Check more response detail at https://docs.sentry.io/api/teams/
       returned: always
       type: complex
     status_code:
       description: Response status code
       returned: always
       type: int
       sample: 200
     url:
       description: requested URL from this module
       returned: always
       type: str
       sample: http://localhost:9000/api/0/organizations/sentry/teams/
"""

from ansible.module_utils.basic import AnsibleModule
from ansible_collections.ridwanbejo.sentry.plugins.module_utils.sentry_api import SentryApi

import requests
import json


def run_module():
    module_args = dict(
        sentry_host=dict(type='str', require=True),
        sentry_token=dict(type='str', require=True),
        organization_slug=dict(type='str', required=True),
        team_slug=dict(type='str', required=False),
        name=dict(type='str', required=False),
        slug=dict(type='str', required=False),
        state=dict(
            default="present", 
            choices=['present', 'absent'],  
            type='str')
    )

    module = AnsibleModule(
        argument_spec=module_args,
        supports_check_mode=True
    )

    if module.check_mode:
        module.exit_json(dict(
            changed=False,
            message='Check mode success!'
        ))

    result = dict(
    )

    sentry_api = SentryApi(module, module.params['sentry_host'], module.params['sentry_token'])

    # a. if state is present then check the existence of team
    if module.params['state'] == "present":

        # a.1. if the team is not exist then create new team
        retrieve_requests = sentry_api.retrieve_team(
            module.params['organization_slug'],
            module.params['team_slug']
        )

        if retrieve_requests['status_code'] == 200:
            result = sentry_api.update_team(
                module.params['organization_slug'],
                module.params['team_slug'],
                module.params['name'],
                module.params['slug'],
            )

            if result['status_code'] != 200:
                module.fail_json(msg=result)

        # a.2. if the team is exist before then update the team
        elif retrieve_requests['status_code'] == 404:

            result = sentry_api.create_team(
                module.params['organization_slug'],
                module.params['name'],
                module.params['slug']
            )

            if result['status_code'] != 201:
                module.fail_json(msg=result)

    # b. if state is absent then delete the team
    elif module.params['state'] == "absent":
        result = sentry_api.delete_team(
            module.params['organization_slug'],
            module.params['team_slug']
        )

        if result['status_code'] != 204:
            module.fail_json(msg=result)

    module.exit_json(**result)


def main():
    run_module()


if __name__ == '__main__':
    main()
